golem-component-service-base/src/repo/component.rs

```diff
@@ -1958,98 +1929,15 @@ impl<Owner: ComponentOwner> ComponentRepo<Owner>
-                   LIMIT 1)
-              INSERT INTO component_versions
-              SELECT prev.component_id, prev.version + 1, prev.size, $2, prev.metadata, prev.component_type, FALSE, prev.object_store_key, prev.transformed_object_store_key, prev.root_package_name, prev.root_package_version FROM prev
-              RETURNING *
-              "#,
-        )
-            .bind(component_id)
-            .bind(Utc::now());
-
-        let new_version = transaction.fetch_one(query).await?.get("version");
-
-        debug!("install_plugin cloned old component version into version {new_version}");
-
-        let old_target = ComponentPluginInstallationRow {
-            component_id,
-            component_version: new_version - 1,
-        };
-        let mut query = self
-            .plugin_installation_queries
-            .get_all(&record.owner, &old_target);
-        let query = query
-            .build_query_as::<PluginInstallationRecord<Owner::PluginOwner, ComponentPluginInstallationTarget>>();
-
-        let existing_installations = transaction.fetch_all(query).await?;
-
-        let new_target = ComponentPluginInstallationRow {
-            component_id,
-            component_version: new_version,
-        };
-
-        let mut new_installations = Vec::new();
-        for installation in existing_installations {
-            let old_id = installation.installation_id;
-            let new_id = Uuid::new_v4();
-            let new_installation = PluginInstallationRecord {
-                installation_id: new_id,
-                target: new_target.clone(),
-                ..installation
-            };
-            new_installations.push(new_installation);
-
-            debug!("install_plugin copying installation {old_id} as {new_id}");
-        }
-        debug!(
-            "install_plugin adding new installation as {}",
-            record.installation_id
-        );
-        new_installations.push(PluginInstallationRecord {
-            target: new_target.clone(),
-            ..record.clone()
-        });
-
-        for installation in new_installations {
-            let mut query = self.plugin_installation_queries.create(&installation);
-            let query = query.build();
-            transaction.execute(query).await?;
-        }
-
-        self.db_pool
-            .with_rw("component", "install_plugin")
-            .commit(transaction)
-            .await?;
-
-        Ok(new_version as u64)
-    }
-
-    async fn uninstall_plugin(
+    async fn apply_plugin_installation_changes(
         &self,
         owner: &<<Owner as ComponentOwner>::PluginOwner as PluginOwner>::Row,
         component_id: &Uuid,
-        plugin_installation_id: &Uuid,
+        actions: &[PluginInstallationRepoAction<Owner>],
     ) -> Result<u64, RepoError> {
         let mut transaction = self
             .db_pool
-            .with_rw("component", "uninstall_plugin")
+            .with_rw("component", "apply_plugin_installation_changes")
```

Comment:
apart from using a transaction, i think this should hold a lock (e.g. SELECT FOR UPDATE) representing the component, otherwise parallel calls will result in mixed applications.

Reply:
This PR is doing the update exactly the same way as before - just applying multiple changes at once (not saying it was not bad before)

Reply:
it was :)

Reply:
i'm fine with postponig the fix, but do not be surprised that this can misbehave until reviewing these